Ensure uploaded images are image files

kbrock opened the pull request:
Consolidated catalog uploads and icon uploads.
Checking image magic numbers to ensure it is a png or jpg

Resolves issue where user could upload a zip bomb

This only addresses

Tested with jpg, gif, png, and zip files

Fryguy commented on app/controllers/mixins/image_validation_mixin.rb:
```diff
@@ -0,0 +1,34 @@
+    private
+
+    # @param file request parameter for a file
+    # @param 
```
I think you cut off your docs?

Fryguy commented on app/controllers/mixins/image_validation_mixin.rb:
```diff
@@ -0,0 +1,34 @@
+    # @param file request parameter for a file
+    # @param 
+    def valid_image_file?(file, type = nil)
+      ext = file.original_filename.split(".").last.downcase
```
Alternatively, `File.extname`

kbrock replied:
`File.extname` keeps the `"."`.
When I changed the code over to that it ended up changing too much.

Deciding to keep it the way it was when I came in

Fryguy commented on app/controllers/mixins/image_validation_mixin.rb:
```diff
@@ -0,0 +1,34 @@
+module Mixins
+  module ImageValidationMixin
+    extend ActiveSupport::Concern
```
I don't think this is needed since this is a normal module without anything extra

Fryguy commented on app/controllers/catalog_controller.rb:
```diff
@@ -540,7 +541,7 @@ def st_upload_image
     elsif params[:upload] && params[:upload][:image] &&
           params[:upload][:image].respond_to?(:read)
       ext = params[:upload][:image].original_filename.split(".").last.downcase
```
This could be moved into the else condition now cause it's not needed in the first condition anymore

kbrock commented:
update:
- fixed `@param` descriptions
- `dropped `Concern`
- use `File.extname`
- moved `ext` to the `else`
- flattened 1 layer of `if`s in the controller

update:
- revert `File.extname()` => `split(".").last`. (it doesn't have a `"."` in the resulting extension.

Fryguy commented:
@kbrock Are there specs for this?  I think the original valid_file? code had specs, which could be moved as a direct test of this mixin.

kbrock commented:
update:
- introduced fixtures directory with images (and mis-typed images)
- testing that catalog does not allow non-image uploads (by file type and not by file name)

Fryguy commented on spec/controllers/catalog_controller_spec.rb:
```diff
@@ -353,13 +353,19 @@
+
+      it "uploads a selected jpg file " do
+        file = fixture_file_upload('spec/fixtures/image.jpg')
+        post :st_upload_image, :params => { :format => :js, :id => @st.id, :upload => {:image => file}, :active_tree => :sandt_tree, :commit => 'Upload' }
+        expect(assigns(:flash_array).first[:message]).to include('Custom Image file "image.jpg" successfully uploaded')
       end
 
       it "displays an error when the selected file is not a png file or .jpg " do
-        file = fixture_file_upload('files/upload_image.txt', 'image/png')
+        file = fixture_file_upload('spec/fixtures/test.txt.png')
```
If I recall correctly, this was intentionally a `.txt` file and not a `.png` file (regardless of the guts of the file) to demonstrate neither .png nor .jpg.

kbrock replied:
Your comment makes a lot of sense.

The upload_spec.rb uses the name `test.txt.png` to show that the controller is validating the file contents and not the extension. Before #8934 `test.txt.png` incorrectly would work. Using that filename shows that we are validating the contents and not the filename.

This PR adds similar validation, so I thought it was a good convention.

Fryguy commented on spec/fixtures/favicon.ico:
ohhhh nevermind - I see now the old code used real files from app/assets, so that's why these are copies

Fryguy commented on spec/controllers/ops_controller/settings/upload_spec.rb:
```diff
@@ -76,21 +67,18 @@
-      controller.params = {:favicon => {:logo => ActionDispatch::Http::UploadedFile.new(:tempfile => File.open(file), :filename => 'test.ico', :type => 'image/vnd.microsoft.icon')}}
-      controller.send(:upload_favicon)
+      controller.params = {:favicon => {:logo => fixture_file_upload('spec/fixtures/favicon.png.ico')}}
+;      controller.send(:upload_favicon)
```
Bonus semicolon?
```suggestion
      controller.send(:upload_favicon)
```

Fryguy commented on spec/controllers/ops_controller/settings/upload_spec.rb:
```diff
@@ -11,58 +11,49 @@
     it 'adds error flash message to @flash_array regarding updating custom logo image with a non-png file' do
-      file = File.join(__dir__, '/data/test.txt.png')
-      controller.params = {:upload => {:logo => ActionDispatch::Http::UploadedFile.new(:tempfile => File.open(file), :filename => 'test.txt.png', :type => 'image/png')}}
+      controller.params = {:upload => {:logo => fixture_file_upload('spec/fixtures/test.txt.png')}}
```
Why was the mime-type not passed to `fixture_file_upload` ?  I'm wondering if that's important for the test, particularly where we're trying to test hacking around file detection like in this case.

kbrock replied:
added mime types to all requests

Fryguy commented:
Skipping backport to `radjabov`, because it is already in the branch.
